### PR TITLE
Ports TG Beam Code and Medigun

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Ports over TG's beam code and the medigun.

Boring stuff first

- ports over TG's cleaned up+datumized beam() code
 - No, this has (almost) nothing to do with laser guns, projectiles, etc, this is a proc for making things "beam", visually from one target to another...like with power gloves, wizard shock, and revenant shock.
 - This will resolve a number of visual issues you see, at times, with revenant, wizard, etc.
 - Powergloves are still 100% functional; yay!

Neato stuff:

- Ports over TG's medigun
 - An item that will heal brute and burn damage, over time, to players that it's connected to--limited to one player at a time.
 - Paradise unique changes: heals broken bones.
 - Nuke ops can purchase it for 15 TC
 - Gamma level medical ERT gets one

medigun

### CODE REVIEW

**Comment by Keilden** [comment]
So Nuke ops can go in TF2 style?

**Comment by Robertha89** [comment]
I was gonna port the medium myself. But beam code. Do not cross the beams.

**Comment by pipextiny1** [comment]
I'm ok with it healing brute, burn and maybe toxin, but bones is a little bit too far for the cost of 15 Tc because a broken bone is one of the worst things that can happen to you in a fire fight and it gives the other person a advantage but if they can negate the effect of a broken bone it won't make it fun for both party's

**Comment by Robertha89** [comment]
Heresy, healing broken bones is a good thing, but thats not even the best part, you can have 2 people with the medigun running around healing the same people, and it will be great.

**Comment by DavidRossii** [comment]
Ubercharging when?

Also I think that's a bit OP to be honest. Healing broken bones mid firefight?
I think perhaps have it on a setting, normal healing and bone reconstruction perhaps? Having you switch between the two.

Also, does the medi-gun have a charge? Similar to laser-guns etc.

**Comment by Robertha89** [comment]
Do remember this, in general, when you break bones it doesnt matter as long as you can still stand at the end, breaking it mid fight overall means little, unless hands, even then. one costs 15, and using it means not using a gun, and instastun bs weapons defeats everything else, so the fact that stun works as they do, as is balances it out in the end.

**Comment by pipextiny1** [comment]
I was more talking about hands and the like, where it breaks and you can't hold a thing most of the broken bones don't matter

File: code/modules/projectiles/medbeam.dm [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	w_class = 3.0
+
+	var/mob/living/current_target
+	var/last_check = 0
+	var/check_delay = 10 //Check los as often as possible, max resolution is SSobj tick though
+	var/max_range = 8
+	var/active = 0
+	var/datum/beam/current_beam = null
+
+	heavy_weapon = 1
```
Comment (TheDZD) [inline code/modules/projectiles/medbeam.dm]: This means that you can't "fire" it while you have something in your other hand, right?

File: code/modules/projectiles/medbeam.dm [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	return
+
+/obj/item/weapon/gun/medbeam/proc/on_beam_tick(var/mob/living/target)
+	target.adjustBruteLoss(-4)
+	target.adjustFireLoss(-4)
+	if(ishuman(target))
+		var/var/mob/living/carbon/human/H = target
+		for(var/obj/item/organ/external/E in H.organs)
+			if(E.mend_fracture())
+				E.perma_injury = 0
```
Comment (TheDZD) [inline code/modules/projectiles/medbeam.dm]: Instant full broken bone healing is probably a bit overboard. Might want to make it only fix bones `if(!getFireLoss() && !getBruteLoss()`, and even then on a prob(10) chance or something.

**Comment by Fox-McCloud** [comment]
E.mend_fracture() kicks in only if the brute damage is below a particular threshold--I can add a probability check though.

File: code/modules/projectiles/medbeam.dm (hunk 1 above)
Comment (Fox-McCloud) [inline code/modules/projectiles/medbeam.dm]: you technically can, but it has consequences--such as the gun flying out of your hands.

**Pull request merged**